src/libpgmoneta/aes.c

```diff
@@ -0,0 +1,228 @@
+   {
+      return 1;
+   }
+
+   return 0;
+}
+
+// [private]
+static int
+aes_encrypt(char* plaintext, unsigned char* key, unsigned char* iv, char** ciphertext, int* ciphertext_length)
+{
+   EVP_CIPHER_CTX* ctx = NULL;
+   int length;
+   size_t size;
+   unsigned char* ct = NULL;
+   int ct_length;
+
+   if (!(ctx = EVP_CIPHER_CTX_new()))
+   {
+      goto error;
+   }
+
+   if (EVP_EncryptInit_ex(ctx, cipher_fp(), NULL, key, iv) != 1)
+   {
+      goto error;
+   }
+
+   size = strlen(plaintext) + EVP_CIPHER_block_size(cipher_fp());
+   ct = malloc(size);
+   memset(ct, 0, size);
+
+   if (EVP_EncryptUpdate(ctx,
+                         ct, &length,
+                         (unsigned char*)plaintext, strlen((char*)plaintext)) != 1)
+   {
+      goto error;
+   }
+
+   ct_length = length;
+
+   if (EVP_EncryptFinal_ex(ctx, ct + length, &length) != 1)
+   {
+      goto error;
+   }
+
+   ct_length += length;
+
+   EVP_CIPHER_CTX_free(ctx);
+
+   *ciphertext = (char*)ct;
+   *ciphertext_length = ct_length;
+
+   return 0;
+
+error:
+   if (ctx)
+   {
+      EVP_CIPHER_CTX_free(ctx);
+   }
+
+   free(ct);
+
+   return 1;
+}
+
+// [private]
+static int
+aes_decrypt(char* ciphertext, int ciphertext_length, unsigned char* key, unsigned char* iv, char** plaintext)
+{
+   EVP_CIPHER_CTX* ctx = NULL;
+   int plaintext_length;
+   int length;
+   size_t size;
+   char* pt = NULL;
+
+   if (!(ctx = EVP_CIPHER_CTX_new()))
+   {
+      goto error;
+   }
+
+   if (EVP_DecryptInit_ex(ctx, cipher_fp(), NULL, key, iv) != 1)
+   {
+      goto error;
+   }
+
+   size = ciphertext_length + EVP_CIPHER_block_size(cipher_fp());
+   pt = malloc(size);
+   memset(pt, 0, size);
+
+   if (EVP_DecryptUpdate(ctx,
+                         (unsigned char*)pt, &length,
+                         (unsigned char*)ciphertext, ciphertext_length) != 1)
+   {
+      goto error;
+   }
+
+   plaintext_length = length;
+
+   if (EVP_DecryptFinal_ex(ctx, (unsigned char*)pt + length, &length) != 1)
+   {
+      goto error;
+   }
+
+   plaintext_length += length;
+
+   EVP_CIPHER_CTX_free(ctx);
+
+   pt[plaintext_length] = 0;
+   *plaintext = pt;
+
+   return 0;
+
+error:
+   if (ctx)
+   {
+      EVP_CIPHER_CTX_free(ctx);
+   }
+
+   free(pt);
+
+   return 1;
+}
+
+// The EVP_aes_256_cbc() is called every time a EVP_CIPHER* is needed,
+// probably because the function is not idempotent.
+// Otherwise we could use EVP_get_cipherbyname(config_term).
+void
+get_cipher()
+{
+   struct configuration* config;
+   config = (struct configuration*)shmem;
+   int len = config->aes_key_length, mode = config->aes_mode;
+   if(len == AES_KEY_LENGTH_256 && mode == AES_MODE_CBC) cipher_fp = &EVP_aes_256_cbc;
+   if(len == AES_KEY_LENGTH_192 && mode == AES_MODE_CBC) cipher_fp = &EVP_aes_192_cbc;
+   if(len == AES_KEY_LENGTH_128 && mode == AES_MODE_CBC) cipher_fp = &EVP_aes_128_cbc;
+   if(len == AES_KEY_LENGTH_256 && mode == AES_MODE_CBC) cipher_fp = &EVP_aes_256_ctr;
```

Comment:
_CTR

Reply:
This part will be fixed with merge 2 configuration parameters into one.

Reply:
Perfect